### ISSUE TITLE
[Numpy][Bug] einsum bug

### ISSUE THREAD

**Issue by sxjscience** [apache/incubator-mxnet#16576]
Example:
```python
import mxnet as mx
import numpy as np
import numpy.testing as npt
from mxnet.gluon import HybridBlock



class MultiNDimBatchDot1(HybridBlock):
    def hybrid_forward(self, F, lhs, rhs):
        """

        Parameters
        ----------
        F
        lhs :
            Shape (N0, N1, T0, C)
        rhs :
            Shape (N0, N1, T1, C)

        Returns
        -------
        ret :
            Shape (N0, N1, T0, T1)
        """
        return F.batch_dot(F.reshape(lhs, (-3, 0, 0)),
                           F.reshape(rhs, (-3, 0, 0)), transpose_b=True).reshape((-4, -1, 8, 0, 0))


class MultiNDimBatchDot2(HybridBlock):
    def hybrid_forward(self, F, lhs, rhs):
        """

        Parameters
        ----------
        F
        lhs :
            Shape (N0, N1, T0, C)
        rhs :
            Shape (N0, N1, T1, C)

        Returns
        -------
        ret :
            Shape (N0, N1, T0, T1)
        """
        return F.np.einsum('abiz,abjz->abij', lhs, rhs)


batch_dot1 = MultiNDimBatchDot1()


batch_dot1.hybridize()

lhs = mx.np.array(np.random.normal(0, 1, (64, 8, 128, 512)), dtype=np.float32, ctx=mx.cpu())
rhs = mx.np.array(np.random.normal(0, 1, (64, 8, 128, 512)), dtype=np.float32, ctx=mx.cpu())
mx.npx.waitall()

gt = np.einsum('abiz,abjz->abij', lhs.asnumpy(), rhs.asnumpy())
out = batch_dot1(lhs.as_nd_ndarray(), rhs.as_nd_ndarray())

npt.assert_allclose(gt, out.asnumpy(), rtol=1E-3, atol=1E-3)


mx.npx.set_np()
batch_dot2 = MultiNDimBatchDot2()
batch_dot2.hybridize()

out2 = batch_dot2(lhs, rhs)
npt.assert_allclose(gt, out2.asnumpy(), rtol=1E-3, atol=1E-3)
```

The CPU results are not correct.

Also, if we use the gpu, it will trigger a segfault:

```python
import mxnet as mx
import numpy as np
import numpy.testing as npt
from mxnet.gluon import HybridBlock



class MultiNDimBatchDot1(HybridBlock):
    def hybrid_forward(self, F, lhs, rhs):
        """

        Parameters
        ----------
        F
        lhs :
            Shape (N0, N1, T0, C)
        rhs :
            Shape (N0, N1, T1, C)

        Returns
        -------
        ret :
            Shape (N0, N1, T0, T1)
        """
        return F.batch_dot(F.reshape(lhs, (-3, 0, 0)),
                           F.reshape(rhs, (-3, 0, 0)), transpose_b=True).reshape((-4, -1, 8, 0, 0))


class MultiNDimBatchDot2(HybridBlock):
    def hybrid_forward(self, F, lhs, rhs):
        """

        Parameters
        ----------
        F
        lhs :
            Shape (N0, N1, T0, C)
        rhs :
            Shape (N0, N1, T1, C)

        Returns
        -------
        ret :
            Shape (N0, N1, T0, T1)
        """
        return F.np.einsum('abiz,abjz->abij', lhs, rhs)


batch_dot1 = MultiNDimBatchDot1()


batch_dot1.hybridize()

lhs = mx.np.array(np.random.normal(0, 1, (64, 8, 128, 512)), dtype=np.float32, ctx=mx.gpu())
rhs = mx.np.array(np.random.normal(0, 1, (64, 8, 128, 512)), dtype=np.float32, ctx=mx.gpu())
mx.npx.waitall()

gt = np.einsum('abiz,abjz->abij', lhs.asnumpy(), rhs.asnumpy())
out = batch_dot1(lhs.as_nd_ndarray(), rhs.as_nd_ndarray())

npt.assert_allclose(gt, out.asnumpy(), rtol=1E-3, atol=1E-3)


mx.npx.set_np()
batch_dot2 = MultiNDimBatchDot2()
batch_dot2.hybridize()

out2 = batch_dot2(lhs, rhs)
npt.assert_allclose(gt, out2.asnumpy(), rtol=1E-3, atol=1E-3)
```

**Comment by sxjscience** [comment]
@hzfan

**Comment by sxjscience** [comment]
I compiled the latest version from source and I'm using g4 machine.

**Comment by hzfan** [comment]
Try build with `-DUSE_INT64_TENSOR_SIZE=ON` (for cmake) or `USE_INT64_TENSOR_SIZE=1` (for make).

**Comment by sxjscience** [comment]
I think the input size has not reached the int64 limit. I tested with a common shape in NLP. Also, I tried to use batch_dot, which could derive the correct result. (See Block1 in the example)

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by hzfan** [comment]
Thanks @sxjscience . The cause may be overflow of intermediate index. I created a fix in #16589.

**Comment by sxjscience** [comment]
I've rerun the script and can confirm that @hzfan's commit has solved the problem. I'll later try to use einsum in the implementation of neural attention.
